File: src/solutions/day17/b.ts
import {readFileSync} from "node:fs";

const contents = readFileSync("input.txt", 'utf-8');
let program = contents.split("\n\n")[1].split(":")[1].trim().split(",").map((v) => Number(v));

const find = (program: number[], answer) => {
    if (program.length === 0) {
        return answer;
    }

    for (let t = 0; t <= 7; t++) {
        let a = answer << 3 | t;
        let b = a % 8;
        b = b ^ 1;
        let c = a >> b;
        b = b ^ 4;
        b = b ^ c;
        if (b % 8 === program[program.length - 1]) {
            let sub = find(program.slice(0, program.length - 1), a);
            if (!sub) continue;
            return sub;
        }
    }
};

console.log("Result: ", find(program, 0));

File: src/solutions/day17/b.py
with open('input.txt') as f:
    lines = f.readlines()

prog = []
for l in lines:
    if l.startswith("Program"):
        prog.extend([int(d) for d in l.split(":")[1].split(",")])

print(prog)
def find(prog, ans):
    if prog == []: return ans

    for t in range(8):
        a = (ans << 3) + t
        b = a % 8
        b = b ^ 1
        c = a >> b
        b = b ^ 4
        b = b ^ c
        if b % 8 == prog[-1]:
            sub = find(prog[:-1], a)
            if sub is None: continue
            return sub

print(find(prog, 0))
        